ipaautomountmap: add support for indirect maps

rjeffman opened the pull request:
Indirect maps were not supported by ansible-freeipa ipaautomountmap. This patch adds support for adding indirect automount maps using the "parent" and "mount" parameters, if the map do not yet exist. An existing map cannot be modified.

The "parent" parameter must match an existing automount map, and the "mount" parameter is required if "parent" is used.

A new test playbook was added to test the feature:

    tests/automount/test_automountmap_indirect.yml

Related:
* RHBZ [#2050176](https://bugzilla.redhat.com/show_bug.cgi?id=2050176)
* RHBZ [#2050158](https://bugzilla.redhat.com/show_bug.cgi?id=2050158)

Fixes #804

varunmylaraiah reviewed (changes requested):
1) The "mount point" option is missing in PR#1075

2) The "automount map indirect" should be successfully added without the parent option.

3) Noticed an wrong error message when trying to add an indirect map

 

  tasks:
  - ipaautomountmap:
      ipaadmin_principal: admin
      ipaadmin_password: <xxxxxxxxxxx>
      automountlocation: 01automuntmap_testlocation
      name: 01_automountmap_testmap
      mount: indirect

TASK [ipaautomountmap] *************************************************************************************************************
task path: /root/02map.yaml:7
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "automountmap_add: 01automuntmap_testlocation: Unknown option: key"}

PLAY RECAP *************************************************************************************************************************
master.ipadomain.test      : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

 

```
CLI Console output:
[root@master ~]# ipa automountmap-add-indirect 01automuntmap_testlocation 01_automountmap_testmap
Mount point: /usr/share/test
---------------------------------------------
Added automount map "01_automountmap_testmap"
---------------------------------------------
  Map: 01_automountmap_testmap
```

varunmylaraiah commented:
from this PR, the expected options
The "automountmap_type" is either "direct" or "indirect, default: direct
The "mountpoint" option should be specified.
The "parentmap" option needs to be included.

rjeffman commented:
@varunmylaraiah I believe I have fixed all this issues, but I found a few more, so the patch is a little more complex.

I've also reviewed the names used in the command-line and fixed both in the module and the README file (a second pair of eyes to check consistency would be really nice).

varunmylaraiah commented:
@rjeffman Thanks for the quick fix. yes, all issues have been fixed. However, I'm still unsure about how to provide a mount point for an indirect map. In CLI, it appears that we cannot add an automount map indirectly without specifying a mount point.


```
[root@master ~]# ipa automountmap-add-indirect
Location: 01automuntmap_testlocation
Map: 01_automountmap_testmap
Mount point:
ipa: ERROR: 'mount' is required
```

rjeffman commented:
So, for the CLI you can use something like:

```
ipa automountmap-add-indirect my_location indirect_map --mount indirect
```

Use `--parentmap` if you want to provide a parent map.

This same command translates to the task:

```yaml
- ipaautomountmap:
    ipaadmin_password: SomeADMINpassword
    location: my_location
    name: indirect_map
    mount: indirect
```

There is no `mount_ponint` paramater as this is used neither on the CLI nor the API. The `mount` parameter is the actual relative mount point (relative to the parent map).

Note that deleting an automount map created with `automountmap-add-indirect` leaves some configuration around (see https://pagure.io/freeipa/issue/9397), and it should be clear if you are using the CLI. The module does the proper cleaning (or should do).

varunmylaraiah reviewed (approved):
LGTM
New test cases for the automount map of type "indirect" have been successfully passed in the downstream

t-woerner commented on README-automountmap.md:
````diff
@@ -54,6 +54,19 @@ Example playbook to ensure presence of an automount map:
       desc: "this is a map for servers in the DMZ"
 ```
 
+Example playbook to ensure an indirect automount map is present:
````
Please explain what an indirect map is and what it is used for.

rjeffman replied:
Fixed. I also added a new example playbook for indirect maps.

t-woerner commented on README-automountmap.md:
````diff
@@ -54,6 +54,21 @@ Example playbook to ensure presence of an automount map:
       desc: "this is a map for servers in the DMZ"
 ```
 
+Automount maps can contain a submount key, which defines a mount location within the map the references another map. On FreeIPA, this is know as an indirect map. An indirect automount map is equivalent to adding a proper automount key to a map, referencyng another map (this second map is the indirect map). Using `parent` and `mount` parameters one can create an indirect automount map with ansible ...
````
There is a missing "n" in "this is know as an indirect map" .

t-woerner replied:
Please rephrase "one can create".

t-woerner reviewed (approved):
LGTM
